Adds initial draft for separate compilation (#96)
cplusplus/SG20: sources/modules/compilation-model/separate-compilation.md

## Module name: Separate Compilation

_Skeleton descriptions are typeset in italic text,_
_so please don't remove these descriptions when editing the topic._

### Overview

_Provides a short natural language abstract of the module’s contents._
_Specifies the different levels of teaching._

This module outlines the issues involved in using separate compilation on multiple program files.

> 

Author: I would merge if I could. Gimme a hint where to do that?i in my fork? elsewhere?

> 

Reviewer: That was resolved in today's meeting.

------------------------------------------------------------------------
Level             Objective
----------------- ------------------------------------------------------

> Foundational:      Awareness of the existence of separate compilation

Reviewer 2: Is 'awareness' here the right term, as the ASBATS of the section require the student to also be able to compile/refactor a program into different translation units.

Reviewer 3: +1

Reviewer: That was resolved in today's meeting.

Main:              Awareness of supporting mechanisms and tools

Advanced:          Awareness of technicalities and tools

------------------------------------------------------------------------

### Motivation

_Why is this important?_
_Why do we want to learn/teach this topic?_

Any non-trivial program will span more than one source file. This makes separate compilation necessary. Separate compilation also leads to quick build times in projects under development.

### Topic introduction

_Very brief introduction to the topic._

> ### Foundational: Using

Reviewer 2: ```suggestion
### Foundational: Awareness of the existence of separate compilation
```

Author: Isn't the term "using" mandated by the template?

Reviewer: That was resolved in today's meeting.

Author: Better caption

#### Background/Required Knowledge

A student: 

> 1. needs to have the text editor skills to split a single-file program into multiple files.

Author: Doesn't hurt to spell it out.

> 1. needs to have the text editor skills to split a single-file program into multiple files.

Reviewer 2: I'd say it actually does: a) this needs is not really related to C++ and b) if we want to consistently specify student requirements with that granularity we need to add many other skills, too that all are not really related to C++.

> 1. needs to have the text editor skills to split a single-file program into multiple files.

Author: I'm fine with removing it. Go ahead and suggest it.

> 1. needs to have the text editor skills to split a single-file program into multiple files.

Reviewer: That was resolved in today's meeting.

2. needs to be able to split function and classes into their declaration and definition.

#### Student outcomes

_A list of things "a student should be able to" after the curriculum._
_The next word should be an action word and testable in an exam._
_Max 5 items._

A student should be able to:

1. Split a single file program into main, one or more auxiliaries, and header files (or modules), all in one directory

> 2. Compile and link a program that is spread over multiple files (again, all in one directory), either on the commandline or with a build system.

Reviewer: Yes, we mentioned in the prerequirements that the student should know about linkage.

3. Describe the functions of the various files involved: source, object, executable.

> #### Points to cover

Author: I don't remember flipping it, but is "caveats" are like footnotes, shouldn't they come after what they cavent?

> #### Points to cover

Author: other participants in the discusion agree with me.

1. Necessity of having function be declared at the locus of use

> 2. Solving this by explicit declaration or by using header files.

Reviewer 2: This point seems to be the answer to the one above, an not a different point to cover? If that's the case, I would suggest to Indent it so the "answer" appears as a sub bullet point.

Reviewer: That was resolved in today's meeting.

Author: fine.

#### Caveats

_This section mentions subtle points to understand, like anything resulting in
implementation-defined, unspecified, or undefined behavior._

1. Explain differences (including in desirability) between having explicit declarations of functions in the main program versus using header files or modules.

> 2. Compilation on a single commandline versus multiple. Resolution order when linking.

Reviewer 2: What is meant by this point? From just the text it's currently for me a bit hard to grasp.

Reviewer: That was resolved in today's meeting.

> 1. Explain differences (including in desirability) between having explicit declarations of functions in the main program versus using header files or modules.
> 2. Compilation on a single commandline versus multiple. Resolution order when linking.
> 3. What needs to be recompiled if only a function is altered? Altered in use syntax versus altered only in semantics.

Reviewer 2: ```suggestion
* Explain differences (including in desirability) between having explicit declarations of functions in the main program versus using header files or modules.
* Compilation on a single command-line versus multiple. Resolution order when linking.
* What needs to be recompiled if only a function is altered? Altered in use syntax versus altered only in semantics.
```

Reviewer: That was resolved in today's meeting.

Author: starred.
not agree with "command-line"

> ### Main: implementing

Reviewer 2: ```suggestion
### Main: Awareness of supporting mechanisms and tools
```

Reviewer: That was resolved in today's meeting.

Author: s/awareness/command/
and bob's your uncle

#### Background/Required Knowledge

* All of the above.

#### Student outcomes

A student should be able to:

> 1. Use compile flags, including `-I` and `-L` search path flags.

Reviewer 2: ```suggestion
1. Use compile flags, including `-I` and `-L` to specify search paths.
```

2. Make header files for their own code, including using header guards.
3. Use include directives for the headers of external libraries.

> 5. Be able to declare `extern` variables.

Reviewer 2: ```suggestion
4. Be able to declare `extern` variables.
```

#### Caveats

> 1. Build systems make things both easier and harder.

Reviewer 2: ```suggestion
* Build systems make things both easier and harder.
```

Reviewer 4: This needs clarification.  A teacher browsing through the document seeing this does not know what is meant by the author.

Author: clarified

#### Points to cover

> 1. Cover the mechanisms for include and library paths to your own code, as well as discovery mechanisms for external libraries.

Reviewer 2: ```suggestion
* Cover the mechanisms for include and library paths to your own code, as well as discovery mechanisms for external libraries.
```

> 

Reviewer 2: From reading the ASBATs, we probably need more points to cover that contain the skills necessary to reach the ASBAT level.

Reviewer: Addressed with the latest commit.

> ### Advanced

Reviewer 2: ```suggestion
### Advanced: Awareness of technicalities and tools
```

Reviewer 2: If there is one, we should put the section name here. However, the one defined in the overview in the beginning seems a bit of (for me) compared to the content that is in the list below.

_These are important topics that are not expected to be covered but provide
guidance where one can continue to investigate this topic in more depth._

> 1. Language inter-operability: the `extern "C"` mechanism.
> 2. Linker conventions including name mangling.
> 3. The `nm` tool for inspection object files and libraries, including de-mangling.
> 4. Understand the issues involved in deciding between include guards and `#pragma once`.

Reviewer 2: ```suggestion
* Language inter-operability: the `extern "C"` mechanism.
* Linker conventions including name mangling.
* The `nm` tool for inspection object files and libraries, including de-mangling.
* Understand the issues involved in deciding between include guards and `#pragma once`.
```

Reviewer: That was resolved in today's meeting.

2. compile and link a program that is spread over multiple files (again, all in one directory), either on the commandline or with a build system.
3. describe the functions of the various files involved: source, object, executable.

#### Points to cover

* Necessity of having function be declared at the locus of use
* Solving this by explicit declaration or by using header files.

#### Caveats

_This section mentions subtle points to understand, like anything resulting in
implementation-defined, unspecified, or undefined behavior._

1. Explain differences (including in desirability) between having explicit declarations of functions in the main program versus using header files or modules.
2. Compilation on a single commandline versus multiple. Resolution order when linking.
3. What needs to be recompiled if only a function is altered? Altered in use syntax versus altered only in semantics.

### Main: implementing

#### Background/Required Knowledge

* All of the above.

#### Student outcomes

A student should be able to:

1. Use compile flags, including `-I` and `-L` to specify search paths.

> 2. Make header files for their own code, including using header guards.

Reviewer 4: I'd prefer it to be explicit that students should be able to reason about translation units (and know what a translation unit is) here, since we are talking about header guards anyways.

Reviewer: That was resolved in today's meeting.

3. Use include directives for the headers of external libraries.

> 4. Be able to declare `extern` variables.

Reviewer 4: Should this really be an explicit goal for a student here?  `extern` variables are rare and bug-prone in practice and making it an explicit goal seems to me as just propagating bad habits.

Reviewer: Maybe we could mention that these are not good practice.

Reviewer: And mention that in some rare cases, we need them.

Reviewer: That was resolved in today's meeting.

### Main: Command of supporting mechanisms and tools

#### Background/Required Knowledge

* All of the above.

#### Student outcomes

A student should be able to:

1. Explain what a translation unit is, and its relation to header files.
2. Use compile flags, including `-I` and `-L` to specify search paths.
3. Make header files for their own code, including using header guards.
4. Use include directives for the headers of external libraries.

#### Caveats

* Build systems make things both easier and harder.
  * Student does not have to set include/library paths.
  * Build systems prevent unnecessary recompilation.
  * Build systems have a separate learning curve.
  * Build systems will pay off when using libraries, less with self-contained student code.

#### Points to cover

* Cover the mechanisms for include and library paths to your own code, as well as discovery mechanisms for external libraries.

### Advanced: Technicalities and tools

_These are important topics that are not expected to be covered but provide
guidance where one can continue to investigate this topic in more depth._

* Language inter-operability: the `extern "C"` mechanism.

> *  Linker conventions including name mangling.

Reviewer 2: ```suggestion
* Linker conventions including name mangling.
```
